Bluetooth: Add dynamic Tx power support on both Host and Controller

rstoica opened the pull request:
This commit targets solving issue #17731 and enabling dynamic Tx control
over the BLE radio controller. The work done here addresses this in the
following way:

- Add public API at the controller interface as
bt_ctrl_set_tx_power(const u32_t power)
- Provide underneath logic on top of the LL_SW arch supporting both
legacy and split versions
- Leverage existent ll_tx_pwr.c unit to implement higher-level
functionality and offer necessary flexibility for both LL_SW legacy
and split variants

LL_SW legacy implementation:
- The main power control logic is implemented at the controller level
in ctrl.c by adding an additional data field to the _radio struct in
order to propagate and sync the radio settings inputed by an user to
the lower radio layer
- Originally, upon init this field, _radio.tx_power_lvl = RADIO_TXP_DEFAULT
and then is dynamically changed by subsequent calls to the public API
- The sync between upstream and downstream layers is provided by the
unified LL_SW API ll_tx_pwr_set, ll_tx_pwr_get and ll_tx_pwr_check
whereas the actual radio control is outsourced to the LLL radio
component via existent radio_tx_power_set.

LL_SW split implementation:
- The power control logic is now outsourced to the vendor implementations
at teh LLL via the lll.h API
	void lll_radio_tx_pwr_set(u32_t power);
	u32_t lll_radio_tx_pwr_get(void);
	u32_t lll_radio_tx_pwr_min_get(void);
	u32_t lll_radio_tx_pwr_max_get(void);
which needs to provide the logic and functionality of the overarching
LL_SW API, i.e., ll_tx_pwr_set, ll_tx_pwr_get and ll_tx_pwr_check.
- The sync is implemented once more over a global variable available
for vendor implementations, i.e. extern u32_t radio_txp.
- This is a bit ugly (I know), but necessary since the LLL runs usually
under hard timing constraints and in ISR contexts. Improvements suggestions
are welcomed, for now this fit the bill and this arch is changing
really fast.
- Again in the LLL_init of the Nordic LLL this variable was set
as radio_txp = RADIO_TXP_DEFAULT;
- The logic is then simple and relies on the low-level radio functionality
to change the Tx power from the prepare_cb of any LLL state, i.e. ADV, SCAN,
MASTER or SLAVE -> this led to the modifications of the form
-	/* TODO: other Tx Power settings */
-	radio_tx_power_set(RADIO_TXP_DEFAULT);
+	radio_tx_power_set(radio_txp);
in the LLL units implementing this functionality

Nordic radio level:
- Utility functions necessary for minimal range validation of the Tx power
at runtime have been implemented across the Nordic SoCs/radios via
	u32_t radio_tx_power_get(void);
	 void radio_tx_power_max_set(void);
	u32_t radio_tx_power_min_get(void);
	u32_t radio_tx_power_max_get(void);
- Upon radio reset the default Kconfig power is set
- Additionally HAL level getters of min powers for the Nordic chips are
implemented according to the supported values provided in the datasheets.

Sample Application:
- A sample application is provied in the form of a beacon advertisement with
dynamic power control
- The dynamic power control is for now supported just over integrated SoCs
at the level of the BLE controller and not propagated at the HCI level as
I believe that is not an applicable use case and there probably vendor specific
commands would be the norm. As a result, dynamic power control over HCI
(qemu support) support is not envisioned with this commit.

The functionality was tested on microbit board and nrf51822 SoCs.

Signed-off-by: Andrei Stoica <stoica.razvan.andrei@gmail.com>

carlescufi commented:
@rstoica thank you very much for this!

I would ask if you could please split this into a few commits instead of having a single big one. Feel free to do this later during the review process once you have addressed any architectural/coding issues. I am thinking perhaps:

- 1x commit for the Legacy LL
- 1x commit for the Split LL
- 1x commit for the sample

rstoica commented:
@carlescufi ok, np. Unfortunately, seems I forgot to run locally the gitlint, so that is already anyways one thing to address. Awaiting other feedback as well.

jhedberg commented:
Don't we want to make this a host-side API? If there's no standard HCI command to set this we could introduce a vendor command, so the sample app would still work e.g. on the nRF91 talking from the M33 to the M4 (nRF52). In general we should try to be conservative with introducing direct controller APIs since those are not portable to any setup where the controller is behind a physical HCI transport.

dbkinder commented on samples/bluetooth/beacon_dp/README.rst:
```diff
@@ -0,0 +1,27 @@
+Overview
+********
+
+A simple application demonstrating the dynamic Tx power control of the LL BLE controller over a BLE Broadcaster node by
```
Let's make this a sentence:
```
This sample application demonstrates the dynamic ...
```

rstoica replied:
Rewrote the sample docs given the changes provided over the HCI commands. Now available under 3bf7dda34231e3991dfd927e4c8a1477163ce797

rstoica commented:
I get your point, and I understand the reasoning behind it. In fact I thought also about it and concluded that is better to start-off to provide the low-level logic and make it publicly available as easier as possible. Hence, my comments in the commit. Tbh, I only started diving into zephyr and BLE 3 days ago, and now that I have a better picture I can take your comments and address the issue on top of the HCI via a VS command (yes, there is no HCI cmd for Tx power). I believe there is even a good pattern to follow with the public address setting.

So, I believe that in order to address your concern the following points need to be implemented:
- add VS command for Tx power setting in hci_vs.h - for example sth like
```
#define BT_HCI_OP_VS_WRITE_BD_ADDR              BT_OP(BT_OGF_VS, 0x0006)
struct bt_hci_cp_vs_write_tx_power {
	u32_t tx_power_lvl;
};
```
- expose existing low-level LL Tx power control via  ```ll_tx_pwr_set``` and corresponding utilities in the hci.c implementation
- modify the current public API ```bt_ctrl_set_tx_power``` to reflect those changes

Anything that I missed ?

What about the public API itself? Would it fine to keep it if these changes are implemented ?

cvinayak commented on include/bluetooth/controller.h:
```diff
@@ -29,6 +29,16 @@ extern "C" {
+ * 
+ *  @return Zero on success or EINVAL error code otherwise.
+ */
+int bt_ctrl_set_tx_power(const u32_t power);
```
As already mentioned by others, a vendor HCI makes more sense for tx power control.

cvinayak commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -83,6 +83,8 @@ u8_t ll_terminate_ind_send(u16_t handle, u8_t reason);
 u8_t ll_rssi_get(u16_t handle, u8_t *rssi);
 u8_t ll_tx_pwr_lvl_get(u16_t handle, u8_t type, s8_t *tx_pwr_lvl);
 void ll_tx_pwr_get(s8_t *min, s8_t *max);
+void ll_tx_pwr_set(const u32_t tx_pwr_lvl);
```
Enhance this with BLE state/role (Adv, Scan, Connection) type, handle per state/role. Could ll_tx_pwr_lvl_get (from HCI) be used as inspiration?

rstoica replied:
@cvinayak thanks for the review and the suggestions! I started applying them, and reached the current solution aiming to solve most of your points:
- expand the LL setter to include LL node role (ADV, SCAN, CONN) and handles as 
```ll_tx_pwr_lvl_set(enum ll_state u16_t handle, u8_t type, s8_t *tx_pwr_lvl);```
- declare in ```ll.h``` the interfaces 
```
enum ll_state {
	ADV,
	SCAN,
	CONN
};

void ll_adv_tx_pwr_lvl_set(u16_t handle, const s8_t tx_pwr_lvl);
void ll_scan_tx_pwr_lvl_set(u16_t handle, const s8_t tx_pwr_lvl);
void ll_conn_tx_pwr_lvl_set(u16_t handle, const s8_t tx_pwr_lvl);
```
that need to be implemented by LL_SW legacy and split versions to provide the dynamic tx control 
- for LL_SW split I have the call path: LL -> ULL -> LLL -> radio -> hal_radio and the logic is now handled at ULL level where the above 3 methods are implemented and at LLL to actually modify the radio Tx in protected BT events
- for LL_SW legacy I am trying to implement now the logic at the controller level decoupling the paths as you mentioned between the roles; however, I saw that atm there are no multiple handles for ADV and SCAN roles; since this path is going to become deprecated I believe than I don't have to add handling of individual ADV/SCAN and just work with the already global advertiser and scanner instances from the radio struct, e.g. https://github.com/zephyrproject-rtos/zephyr/blob/fca32e41e608dd4751df2e7e664f1cce70b37998/subsys/bluetooth/controller/ll_sw/ctrl.c#L195 and 
https://github.com/zephyrproject-rtos/zephyr/blob/fca32e41e608dd4751df2e7e664f1cce70b37998/subsys/bluetooth/controller/ll_sw/ctrl.c#L196
would this be acceptable ?
- the rest of the work then is to simply create a VS HCI command and implement some examples and tests.

Any comments from your side ?

cvinayak replied:
I agree with most. Thumbs up towards the feature enhancements you are doing :-)

Instead of three new interface just change:
```u8_t ll_tx_pwr_lvl_get(u16_t handle, u8_t type, s8_t *tx_pwr_lvl);``` to ```u8_t ll_tx_pwr_lvl_get(enum ll_state state, u16_t handle, u8_t type, s8_t *tx_pwr_lvl);```
In the hci.c where it is used, pass CONN as the state, and in new vendor HCI interface pass state as supplied by application.

and add:
```u8_t ll_tx_pwr_lvl_set(enum ll_state state, u16_t handle, s8_t tx_pwr_lvl);``` for dynamic control.

For the legacy controller, ignore the handle supplied and directly reference the singular advertiser and scanner structs.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -269,6 +269,9 @@ static struct {
 	struct connection *conn_upd;
+
+	/* Dynamic Tx Power allocation */
+	u32_t tx_power_lvl;
```
Name this ```tx_pwr_lvl_default```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -515,6 +518,9 @@ u32_t radio_init(void *hf_clock, u8_t sca, void *entropy,
 	_radio.fc_ena = 1U;
 
+	/* Set the initial Tx power to the default one set via Kconfig */
+	_radio.tx_power_lvl = RADIO_TXP_DEFAULT;
```
```tx_pwr_lvl_default =```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -6584,10 +6590,32 @@ static u32_t access_addr_get(void)
 static void adv_scan_conn_configure(void)
```
Delete this function, move content to callers. Use state/role values of ```tx_pwr_lvl```. 

```adv->tx_pwr_lvl```
```scan->tx_pwr_lvl```
```conn->tx_pwr_lvl```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -6584,10 +6590,32 @@ static u32_t access_addr_get(void)
 	radio_isr_set(isr, NULL);
 }
 
+void ll_tx_pwr_get(s8_t *min, s8_t *max)
```
Why are you duplicating this, its already in ll_tx_pwr.c ?

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -6584,10 +6590,32 @@ static u32_t access_addr_get(void)
+	*max = (s8_t)radio_tx_power_max_get();
+}
+
+void ll_tx_pwr_set(const u32_t tx_power_lvl) 
```
Keep the interface in ll_tx_pwr.c

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -6584,10 +6590,32 @@ static u32_t access_addr_get(void)
+void ll_tx_pwr_set(const u32_t tx_power_lvl) 
+{
+	_radio.tx_power_lvl = tx_power_lvl;
+	radio_tx_power_set(_radio.tx_power_lvl);
```
h/w registers should only be accessed in a controlled fashion, inside respect state/role's reserved time space (BLE radio events)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -6584,10 +6590,32 @@ static u32_t access_addr_get(void)
+	radio_tx_power_set(_radio.tx_power_lvl);
+}
+
+u8_t ll_tx_pwr_check(const u32_t tx_power_lvl)
```
Dont see the need for this interface, give a set and get interface applications can do the check

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -12354,12 +12382,13 @@ u8_t ll_tx_pwr_lvl_get(u16_t handle, u8_t type, s8_t *tx_pwr_lvl)
-	 *       updates is implemented.
-	 */
-	*tx_pwr_lvl = RADIO_TXP_DEFAULT;
+	if (type) {
```
this change approved.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -12354,12 +12382,13 @@ u8_t ll_tx_pwr_lvl_get(u16_t handle, u8_t type, s8_t *tx_pwr_lvl)
+		*tx_pwr_lvl = (s8_t)radio_tx_power_max_get();
+	} else {
+		/* Current level is requested */
+		*tx_pwr_lvl = (s8_t)_radio.tx_power_lvl;
```
Use ```conn->tx_pwr_lvl```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_tx_pwr.c:
```diff
@@ -31,22 +33,58 @@ u8_t ll_tx_pwr_lvl_get(u16_t handle, u8_t type, s8_t *tx_pwr_lvl)
+		*tx_pwr_lvl = (s8_t)lll_radio_tx_pwr_max_get();
+	} else {
+		/* Current level is requested */
+		*tx_pwr_lvl = (s8_t)lll_radio_tx_pwr_get();
```
Define lll_conn_tx_pwr_get

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -114,6 +114,7 @@ void radio_reset(void)
 			    << RADIO_POWER_POWER_Pos) & RADIO_POWER_POWER_Msk);
 	nrf_radio_power_set((RADIO_POWER_POWER_Enabled << RADIO_POWER_POWER_Pos)
 			    & RADIO_POWER_POWER_Msk);
+	radio_tx_power_set(RADIO_TXP_DEFAULT);
```
No!

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -137,6 +138,11 @@ void radio_phy_set(u8_t phy, u8_t flags)
 #endif /* CONFIG_BT_CTLR_RADIO_ENABLE_FAST */
 }
 
+u32_t radio_tx_power_get(void)
```
You dont need this if the check interface is also not defined.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:
```diff
@@ -437,6 +440,27 @@ u32_t lll_radio_is_idle(void)
+	return radio_tx_power_get();
+}
+
+u32_t radio_txp;
```
Global variable not here. Start using bottom up, example, declare in ```struct lll_adv``` ```struct lll_conn``` etc. use this in individual states/role, then move up defining a private interface that satisfies both downstream and upstream (HCI, Application API) modules.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -142,8 +142,7 @@ static int prepare_cb(struct lll_prepare_param *prepare_param)
 	radio_reset();
-	/* TODO: other Tx Power settings */
-	radio_tx_power_set(RADIO_TXP_DEFAULT);
+	radio_tx_power_set(radio_txp);
```
```lll->tx_pwr_lvl```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_master.c:
```diff
@@ -138,8 +138,8 @@ static int prepare_cb(struct lll_prepare_param *prepare_param)
 	radio_reset();
-	/* TODO: other Tx Power settings */
-	radio_tx_power_set(RADIO_TXP_DEFAULT);
+	radio_tx_power_set(radio_txp);
```
```lll->tx_pwr_lvl```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -145,8 +145,7 @@ static int prepare_cb(struct lll_prepare_param *prepare_param)
 	radio_reset();
-	/* TODO: other Tx Power settings */
-	radio_tx_power_set(RADIO_TXP_DEFAULT);
+	radio_tx_power_set(radio_txp);
```
```lll->tx_pwr_lvl```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_slave.c:
```diff
@@ -159,8 +159,7 @@ static int prepare_cb(struct lll_prepare_param *prepare_param)
 	radio_reset();
-	/* TODO: other Tx Power settings */
-	radio_tx_power_set(RADIO_TXP_DEFAULT);
+	radio_tx_power_set(radio_txp);
```
```lll->tx_pwr_lvl```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.h:
```diff
@@ -14,7 +14,10 @@ void radio_setup(void);
 void radio_tx_power_set(u32_t power);
+u32_t radio_tx_power_get(void);
 void radio_tx_power_max_set(void);
+u32_t radio_tx_power_min_get(void);
```
Accept the lll->radio->hal_radio call path for radio_tx_power_min_get and radio_tx_power_max_get.

cvinayak reviewed (changes requested):
My change requests are towards state/role based tx power level control, HCI already has cmd to get per connection tx pwr level.

Add a Kconfig to conditionally compile out the dynamic tx pwr lvl control.

mtpr-ot reviewed (dismissed):
I agree with @cvinayak's comments - especially that it must be possible to set TX power for individual roles/states, but also individual connections.

joerchan commented:
> Don't we want to make this a host-side API? If there's no standard HCI command to set this we could introduce a vendor command, so the sample app would still work e.g. on the nRF91 talking from the M33 to the M4 (nRF52). In general we should try to be conservative with introducing direct controller APIs since those are not portable to any setup where the controller is behind a physical HCI transport.

I agree to this point. I think we should go through the HCI interface to do this. In order to do that we might want to add these commands to the zephyr HCI extensions.

rstoica commented:
> > Don't we want to make this a host-side API? If there's no standard HCI command to set this we could introduce a vendor command, so the sample app would still work e.g. on the nRF91 talking from the M33 to the M4 (nRF52). In general we should try to be conservative with introducing direct controller APIs since those are not portable to any setup where the controller is behind a physical HCI transport.
> 
> I agree to this point. I think we should go through the HCI interface to do this. In order to do that we might want to add these commands to the zephyr HCI extensions.

@joerchan just to clarify the nomenclature: by "Zephyr HCI extensions" you actually mean HCI vendor specific commands, i.e. the ones in `hci_vs.h`, correct ?

rstoica commented:
> > > Don't we want to make this a host-side API? If there's no standard HCI command to set this we could introduce a vendor command, so the sample app would still work e.g. on the nRF91 talking from the M33 to the M4 (nRF52). In general we should try to be conservative with introducing direct controller APIs since those are not portable to any setup where the controller is behind a physical HCI transport.
> > 
> > 
> > I agree to this point. I think we should go through the HCI interface to do this. In order to do that we might want to add these commands to the zephyr HCI extensions.
> 
> @joerchan just to clarify the nomenclature: by "Zephyr HCI extensions" you actually mean HCI vendor specific commands, i.e. the ones in `hci_vs.h`, correct ?

Nvm, got you now -> you were refering to them being added to the HCI VS supported functionality under `CONFIG_BT_HCI_VS_EXT`

joerchan commented:
> > > > Don't we want to make this a host-side API? If there's no standard HCI command to set this we could introduce a vendor command, so the sample app would still work e.g. on the nRF91 talking from the M33 to the M4 (nRF52). In general we should try to be conservative with introducing direct controller APIs since those are not portable to any setup where the controller is behind a physical HCI transport.
> > > 
> > > 
> > > I agree to this point. I think we should go through the HCI interface to do this. In order to do that we might want to add these commands to the zephyr HCI extensions.
> > 
> > 
> > @joerchan just to clarify the nomenclature: by "Zephyr HCI extensions" you actually mean HCI vendor specific commands, i.e. the ones in `hci_vs.h`, correct ?
> 
> Nvm, got you now -> you were refering to them being added to the HCI VS supported functionality under `CONFIG_BT_HCI_VS_EXT`

@rstoica I'm specifically referring to these: https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/reference/bluetooth/hci.txt
It would be easier if it was defined as extensions here.

rstoica committed:
Bluetooth: controller: enable dynamic TXP over LL_SPLIT arch (#17731)

This commit targets solving issue #17731 over the LL_SW_SPLIT
arch of the BLE stack in Zephyr. This functionality is exposed
to the user as HCI Zephyr Command extensions

- BT_HCI_OP_VS_WRITE_TX_POWER_LEVEL
- BT_HCI_OP_VS_READ_TX_POWER_LEVEL

which enable Tx power read/write operations within BLE radio events
on a per role/connection basis.

The functionality is enabled upon the Kconfig advanced configuration
triggered by

- BT_CTLR_TX_PWR_DYNAMIC_CONTROL

depending on the enablement of Zephyr HCI vendor-specific command
extensions.

Necessary low-level radio HAL functionality and power definitions
are also supplied to address the high-level functionality of
controlling the Tx power.

Signed-off-by: Andrei Stoica <stoica.razvan.andrei@gmail.com>

rstoica committed:
Bluetooth: controller: enable dynamic TXP over LEGACY arch (#17731)

This commit targets solving issue #17731 over the Nordic LL LEGACY
arch of the BLE stack in Zephyr. This functionality is exposed
to the user as HCI Zephyr Command extensions

- BT_HCI_OP_VS_WRITE_TX_POWER_LEVEL
- BT_HCI_OP_VS_READ_TX_POWER_LEVEL

which enable Tx power read/write operations within BLE radio events
on a per role/connection basis.

The functionality is enabled upon the Kconfig advanced configuration
triggered by

- BT_CTLR_TX_PWR_DYNAMIC_CONTROL

depending on the enablement of Zephyr HCI vendor-specific command
extensions.

Signed-off-by: Andrei Stoica <stoica.razvan.andrei@gmail.com>

rstoica committed:
docs: Bluetooth: HCI_VS: Updated docs on VS HCI commands

This commit complements documentation of Zephyr HCI VS
commands. In particular, it documents the newly introduced
commands

- BT_HCI_OP_VS_WRITE_TX_POWER_LEVEL
- BT_HCI_OP_VS_READ_TX_POWER_LEVEL

These provide HCI-level control over the Tx power of the
BLE radio on a per role/connection basis.

The functionality is enabled upon the Kconfig advanced configuration
triggered by

- BT_CTLR_TX_PWR_DYNAMIC_CONTROL

depending on the enablement of Zephyr HCI vendor-specific command
extensions.

Signed-off-by: Andrei Stoica <stoica.razvan.andrei@gmail.com>

rstoica committed:
Bluetooth: samples: Added dynamic Tx power ctrl sample (#17731)

This commit provides a sample dummy application demonstrating
the usage of the added dynamic Tx power control over the HCI
commands and HCI interfaces.

Signed-off-by: Andrei Stoica <stoica.razvan.andrei@gmail.com>

jhedberg commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,307 @@
+}
+
+
+static void set_tx_power(u8_t state,  u16_t handle, s8_t tx_pwr_lvl)
```
I'm fine doing this incrementally, i.e. host APIs in follow-up PR. Btw, how does the adv power API fit together with extended advertising, i.e. do we need per-advertising instance power settings?

joerchan replied:
Ok. let's do it by increments then.

The HCI commands here allow for per-advertising (advertising sets I think) instance power settings. So this should work fine for advertising exensions API as well.

rstoica replied:
I can extend the host API once the PR passes and everyone is on board with it.

rstoica commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -143,8 +150,11 @@ static int prepare_cb(struct lll_prepare_param *prepare_param)
 	radio_reset();
-	/* TODO: other Tx Power settings */
+#if defined(CONFIG_BT_TX_PWR_DYNAMIC_CTRL)
+	radio_tx_power_set((u32_t)((s32_t)lll->tx_pwr_lvl));
```
I think it can be a bit improved by casting directly to `u32_t` as the `s8_t` to `s32_t` extension should happen automatically. The reason I use `s8_t` is very simple. The memory footprint scales with the no. of connections allowed and each connection would then have 3 extra bytes which are unused and unnecessary.

On the other hand, on the radio low-level side the inputs modify directly the 32bit registers, hence the `u32_t` type for the radio Tx power set interface.

joerchan replied:
If we use s8_t as the type, then it should be that for all functions accepting this parameter. The cast to u32_t for the HW mmio can be done inside that particular function.
My 2 cents :)

rstoica replied:
@cvinayak would you agree for the HAL setter so I could rework this ?

cvinayak replied:
yes. Have it as separate commit.

rstoica replied:
Done in 6716117a10

rstoica commented on include/bluetooth/hci_vs.h:
```diff
@@ -129,6 +129,35 @@ struct bt_hci_cp_vs_set_scan_req_reports {
+	u8_t ll_state;
+	u16_t handle;
+	s8_t tx_power_level;
+} __packed;
```
ok, np will reformat

rstoica commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,307 @@
+	if (err) {
+		printk("TXP GET HCI err: %d\n", err);
+		net_buf_unref(buf);
+		net_buf_unref(rsp);
```
you're right! thanks, missed the unref in the `bt_hci_cmd_send_sync` variant! will change!

rstoica commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,307 @@
+const s8_t txp[DEVICE_BEACON_TXPOWER_NUM] = {4, 0, -4, -8, -12, -20, -24, -30};
+const struct bt_le_adv_param *param = BT_LE_ADV_PARAM
+		(BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_USE_NAME,
+		0x0020, 0x0020);
```
static - changed!

rstoica commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,307 @@
+	if (!buf) {
+		err = -ENOBUFS;
+		printk("Err: %d\n", err);
+		net_buf_unref(buf);
```
Hahaha, sometimes I am a funny guy... yeah, sorry, changed!

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -99,6 +99,13 @@ int lll_adv_reset(void)
 }
 
+#if defined(CONFIG_BT_TX_PWR_DYNAMIC_CTRL)
+void lll_adv_tx_pwr_reset(struct lll_adv *lll)
```
You dont need these if you are only reseting the value on state/role teardown. Remove every where, only assign when state/role is instantiated.

rstoica replied:
Implemented TXP defaults at instantiation/setup procedures level.

Advertiser state: https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/a0f3d32567b51755d3f2e8869dfe1e4c3f487a2d#diff-7dd8f83650705279a5ab78ea1e56798dR450

Advertiser state - connection provision: https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/a0f3d32567b51755d3f2e8869dfe1e4c3f487a2d#diff-7dd8f83650705279a5ab78ea1e56798dR604

Scanner state: https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/a0f3d32567b51755d3f2e8869dfe1e4c3f487a2d#diff-5fde7b773bf177bd42dddfd1715205a1R107

Master role: https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/a0f3d32567b51755d3f2e8869dfe1e4c3f487a2d#diff-50a16934d1f6f0c7540768f7c36cffefR147

Slave role: https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/a0f3d32567b51755d3f2e8869dfe1e4c3f487a2d#diff-1cfc208e525b4f77177d8a72616b84a7R172

rstoica commented on subsys/bluetooth/controller/ll_sw/ll_tx_pwr.c:
```diff
@@ -9,6 +9,7 @@
 #include <toolchain.h>
 #include <soc.h>
 #include <bluetooth/hci.h>
+#include <bluetooth/controller.h>
```
Dead code. Removed!

rstoica commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,307 @@
+		printk("TXP SET HCI err: %d\n", err);
+	}
+
+	net_buf_unref(buf);
```
Fixed in https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/15b6ef56996c74281b078992eaa1afe67025b156#diff-b9bd630d9245e9d01d535cbaa393a233R96

rstoica commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,307 @@
+static struct bt_conn *default_conn;
+static u16_t default_conn_handle;
+
+static u8_t connflag;
```
Done - https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/15b6ef56996c74281b078992eaa1afe67025b156#diff-b9bd630d9245e9d01d535cbaa393a233R171

rstoica commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -7,6 +7,12 @@
 
 #define LL_VERSION_NUMBER BT_HCI_VERSION_5_1
 
+enum ll_state {
```
Done - https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/a0f3d32567b51755d3f2e8869dfe1e4c3f487a2d#diff-65c5dce42a06bd585801995fc7a03ad0R10

rstoica commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -515,6 +523,12 @@ u32_t radio_init(void *hf_clock, u8_t sca, void *entropy,
 	 */
 	_radio.fc_ena = 1U;
 
+	/* Set the initial Tx power to the default one set via Kconfig */
```
Done - https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/9b2ff3fa0e8758f4cd178faece8c881bb95b6183#diff-cfe3bbb9af7efaaa75e0a9c4dc4da2a3R661

rstoica commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -565,6 +579,11 @@ void ll_reset(void)
 	_radio.packet_release_first = 0U;
 	_radio.packet_release_last = 0U;
 
+#if defined(CONFIG_BT_TX_PWR_DYNAMIC_CTRL)
```
Done - https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/9b2ff3fa0e8758f4cd178faece8c881bb95b6183#diff-cfe3bbb9af7efaaa75e0a9c4dc4da2a3R661

rstoica commented on subsys/bluetooth/controller/ll_sw/ctrl_internal.h:
```diff
@@ -328,6 +328,10 @@ struct connection {
 #endif /* CONFIG_BT_CTLR_CONN_RSSI */
+
+#if defined(CONFIG_BT_TX_PWR_DYNAMIC_CTRL)
+	u32_t tx_pwr_lvl;
```
Fixed - https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/9b2ff3fa0e8758f4cd178faece8c881bb95b6183#diff-7e50ac9f07d9481702d932ffb428ba70R333

rstoica commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,307 @@
+	cp->tx_power_level = tx_pwr_lvl;
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_VS_WRITE_TX_POWER_LEVEL,
+			buf, NULL);
```
Now fixed - see for instance https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/1eddc7d08df80637ae70155480d952b8a138658d#diff-b9bd630d9245e9d01d535cbaa393a233R83

rstoica commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,307 @@
+		*tx_pwr_lvl = rp->tx_power_level;
+	}
+
+	net_buf_unref(buf);
```
Fixed - https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/1eddc7d08df80637ae70155480d952b8a138658d#diff-b9bd630d9245e9d01d535cbaa393a233R134

Also same for the others

rstoica commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -201,6 +201,14 @@ config BT_CTLR_TX_PWR_MINUS_40
 
 endchoice
 
+config BT_TX_PWR_DYNAMIC_CTRL
```
Done - https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/1535a38a863cb7f0c4b1c0cef140e6282928b5e9#diff-fdf68fcfb91aad56ca9d8b0a627c1ab4R204

rstoica commented on samples/bluetooth/hci_pwr_ctrl/README.rst:
```diff
@@ -0,0 +1,30 @@
+
+* BlueZ running on the host, or
+* A board with BLE support
+* A central device & monitor (e.g. nRF Connect) to check the RSSI values obtained from the peripheral.
```
Done - https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/1eddc7d08df80637ae70155480d952b8a138658d#diff-51497db637fdff050fede888928d2b39R27

joerchan reviewed (commented):
I made some of these comments before the slack discussion.

I think we need to discuss this next week when @jhedberg @carlescufi are back.
@rstoica Maybe we could take this on slack on monday?

rstoica commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,292 @@
+							  addr, sizeof(addr));
+			printk("Connected via connection (%d) at %s\n",
+			       default_conn_handle, addr);
+			set_tx_power(2, default_conn_handle, 4);
```
Sure

rstoica replied:
[Done](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/5e93cb2ee03a927766c172edb1950e774b6bf16c#diff-b9bd630d9245e9d01d535cbaa393a233R168)

rstoica commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,292 @@
+
+	bt_conn_cb_register(&conn_callbacks);
+
+	k_sleep(K_SECONDS(5));
```
It is just a convenience way to test / display the fact that default Kconfig power is properly set up upon BLE stack setup and initial advertising without any request to change Tx power. Having the 5 seconds delay or so one can see that indeed the level set up for `RADIO_TXP_DEFAULT` is properly used in the very beginning upon reset.

joerchan replied:
Ok. You might want to comment that though. As it not obvious that that is what is happening :)

rstoica replied:
Addressed this both inline at the [code level](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/5e93cb2ee03a927766c172edb1950e774b6bf16c#diff-b9bd630d9245e9d01d535cbaa393a233R301) and in the [documentation](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/5e93cb2ee03a927766c172edb1950e774b6bf16c#diff-51497db637fdff050fede888928d2b39R9) of the sample itself.

rstoica commented on samples/bluetooth/hci_pwr_ctrl/README.rst:
```diff
@@ -0,0 +1,35 @@
+Overview
+********
+
+This sample application demonstrates the dynamic Tx power control over HCI RAM
```
Yes, you are right, actually works over USB as tested... basically this functionality is provided over the HCI regardless of the transport layer. Will modify!

rstoica replied:
Reworked the sample [documentation](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/5e93cb2ee03a927766c172edb1950e774b6bf16c#diff-51497db637fdff050fede888928d2b39R9) to avoid confusion.

joerchan commented:
@rstoica 
Really sorry but I only now noticed that we actually have this for our SoftDevices.
It's documented here in case you are interested: https://infocenter.nordicsemi.com/topic/com.nordic.infocenter.s132.api.v7.0.1/group___b_l_e___g_a_p___f_u_n_c_t_i_o_n_s.html?cp=3_5_2_1_2_1_3_41#gad10e80ba20ddab42d1753fd1be2d8c26

This is an API that exists alongside with Advertising Extension.
I had someone who from the SoftDevice team look at this PR when you first posted it, and they did not raise any issues. So this approach seems fine.

I'm wondering if we could align the HCI extension to be compatible at the SD HCI.
I have posted our HCI version of this command here:
```
struct {
	uint8_t role; /* BLE_GAP_TX_POWER_ROLE_ */
	uint16_t handle; /* adv or conn handle */
	int8_t tx_power; /* NRF_OUTPUT_POWER_ */
} hci_set_tx_power;

  BLE_GAP_TX_POWER_ROLE_ADV       = 1,           /**< Advertiser role. */
  BLE_GAP_TX_POWER_ROLE_SCAN_INIT = 2,           /**< Scanner and initiator role. */
  BLE_GAP_TX_POWER_ROLE_CONN      = 3,           /**< Connection role. */

  NRF_OUTPUT_POWER_PLUS_8DBM   =  8,   /**< Output power set to  8dBm  */
  NRF_OUTPUT_POWER_PLUS_7DBM   =  7,   /**< Output power set to  7dBm  */
  NRF_OUTPUT_POWER_PLUS_6DBM   =  6,   /**< Output power set to  6dBm  */
  NRF_OUTPUT_POWER_PLUS_5DBM   =  5,   /**< Output power set to  5dBm  */
  NRF_OUTPUT_POWER_PLUS_4DBM   =  4,   /**< Output power set to  4dBm  */
  NRF_OUTPUT_POWER_PLUS_3DBM   =  3,   /**< Output power set to  3dBm  */
  NRF_OUTPUT_POWER_PLUS_2DBM   =  2,   /**< Output power set to  2dBm  */
  NRF_OUTPUT_POWER_0DBM        =  0,   /**< Output power set to  0dBm (default) */
  NRF_OUTPUT_POWER_MINUS_4DBM  = -4,   /**< Output power set to -4dBm  */
  NRF_OUTPUT_POWER_MINUS_8DBM  = -8,   /**< Output power set to -8dBm  */
  NRF_OUTPUT_POWER_MINUS_12DBM = -12,  /**< Output power set to -12dBm */
  NRF_OUTPUT_POWER_MINUS_16DBM = -16,  /**< Output power set to -16dBm */
  NRF_OUTPUT_POWER_MINUS_20DBM = -20,  /**< Output power set to -20dBm */
  NRF_OUTPUT_POWER_MINUS_40DBM = -40   /**< Output power set to -40dBm */
```

rstoica commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -102,6 +103,10 @@ u8_t ll_scan_enable(u8_t enable)
 	}
 #endif
 
+#if defined(CONFIG_BT_CTRL_TX_PWR_DYNAMIC_CTRL)
```
This needs to be moved to `ull_scan_enable` below!

rstoica replied:
Done [here](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-5fde7b773bf177bd42dddfd1715205a1R178)

rstoica commented:
> @rstoica
> Really sorry but I only now noticed that we actually have this for our SoftDevices.
> It's documented here in case you are interested: https://infocenter.nordicsemi.com/topic/com.nordic.infocenter.s132.api.v7.0.1/group___b_l_e___g_a_p___f_u_n_c_t_i_o_n_s.html?cp=3_5_2_1_2_1_3_41#gad10e80ba20ddab42d1753fd1be2d8c26
> 
> This is an API that exists alongside with Advertising Extension.
> I had someone who from the SoftDevice team look at this PR when you first posted it, and they did not raise any issues. So this approach seems fine.
> 
> I'm wondering if we could align the HCI extension to be compatible at the SD HCI.
> I have posted our HCI version of this command here:
> 
> ```
> struct {
> 	uint8_t role; /* BLE_GAP_TX_POWER_ROLE_ */
> 	uint16_t handle; /* adv or conn handle */
> 	int8_t tx_power; /* NRF_OUTPUT_POWER_ */
> } hci_set_tx_power;
> 
>   BLE_GAP_TX_POWER_ROLE_ADV       = 1,           /**< Advertiser role. */
>   BLE_GAP_TX_POWER_ROLE_SCAN_INIT = 2,           /**< Scanner and initiator role. */
>   BLE_GAP_TX_POWER_ROLE_CONN      = 3,           /**< Connection role. */
> 
>   NRF_OUTPUT_POWER_PLUS_8DBM   =  8,   /**< Output power set to  8dBm  */
>   NRF_OUTPUT_POWER_PLUS_7DBM   =  7,   /**< Output power set to  7dBm  */
>   NRF_OUTPUT_POWER_PLUS_6DBM   =  6,   /**< Output power set to  6dBm  */
>   NRF_OUTPUT_POWER_PLUS_5DBM   =  5,   /**< Output power set to  5dBm  */
>   NRF_OUTPUT_POWER_PLUS_4DBM   =  4,   /**< Output power set to  4dBm  */
>   NRF_OUTPUT_POWER_PLUS_3DBM   =  3,   /**< Output power set to  3dBm  */
>   NRF_OUTPUT_POWER_PLUS_2DBM   =  2,   /**< Output power set to  2dBm  */
>   NRF_OUTPUT_POWER_0DBM        =  0,   /**< Output power set to  0dBm (default) */
>   NRF_OUTPUT_POWER_MINUS_4DBM  = -4,   /**< Output power set to -4dBm  */
>   NRF_OUTPUT_POWER_MINUS_8DBM  = -8,   /**< Output power set to -8dBm  */
>   NRF_OUTPUT_POWER_MINUS_12DBM = -12,  /**< Output power set to -12dBm */
>   NRF_OUTPUT_POWER_MINUS_16DBM = -16,  /**< Output power set to -16dBm */
>   NRF_OUTPUT_POWER_MINUS_20DBM = -20,  /**< Output power set to -20dBm */
>   NRF_OUTPUT_POWER_MINUS_40DBM = -40   /**< Output power set to -40dBm */
> ```

@joerchan thanks for the reference. To align the 2 HCI commands would basically imply the +1 shift in the role mapping [here](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/1535a38a863cb7f0c4b1c0cef140e6282928b5e9#diff-65c5dce42a06bd585801995fc7a03ad0R10) and [here](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/1535a38a863cb7f0c4b1c0cef140e6282928b5e9#diff-d32c7d39d131dcc56d85c8b8b7d5b328R132). The types and serialization of the arguments seems *ready* to me. Shall the naming convention also be followed? I believe so, especially for the role part, whereas for the Tx power level I think your previous suggestion, i.e. `Tx_Power_level` is better suited following the Core BT spec.

I am not sure about the power levels. Where would they go? In `hci_vs.h` seems a bit forced since they are vendor specific. Would everyone be supposed to add their own ones there ?

carlescufi commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -800,6 +800,140 @@ When the Set_Scan_Request_Reports command has completed, a Command Complete
+	+--------------------+--------------------------------------+
+	| Value              | Parameter Description                |
+	+--------------------+--------------------------------------+
+	| -128 <= N <= 127   | The desired Tx Power Level to be     |
```
Shouldn't this be `-128 <= N <= 20`? That's what the spec has for the `LE Read Transmit Power` command

rstoica replied:
Thanks for the improvement suggestion with the 0x7F (127) [level](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-d32c7d39d131dcc56d85c8b8b7d5b328R136). I implemented that [here](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-611c010769adb02c1f8eb406479efa93R128) and also addressed the documentation following the spec [here](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/6012d2b2ca1f5515ee0ae6dfc760301bfc4b0faa#diff-9c37bb9bf44e4a2035f47bde14315da6R878). As per spec I formatted `-127 <= N <= 126` and [clarified](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/6012d2b2ca1f5515ee0ae6dfc760301bfc4b0faa#diff-9c37bb9bf44e4a2035f47bde14315da6R884-R892) the flooring operation inherently [performed](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-611c010769adb02c1f8eb406479efa93R138) to address varying power levels over different controllers.

carlescufi commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -800,6 +800,140 @@ When the Set_Scan_Request_Reports command has completed, a Command Complete
+	|                    | modify                               |
+	+--------------------+--------------------------------------+
+
+	Tx Power level:                                Size: 2 Octets
```
This must be 1 octet instead

rstoica replied:
Fixed everywhere in [HCI VS docs](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/6012d2b2ca1f5515ee0ae6dfc760301bfc4b0faa#diff-9c37bb9bf44e4a2035f47bde14315da6R878)

carlescufi commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -800,6 +800,140 @@ When the Set_Scan_Request_Reports command has completed, a Command Complete
+| Write_Tx_Power_Level     | 0x00E | LL_State,          | Status,            |
+|                          |       | Handle,            | LL_State,          |
+|                          |       | Tx Power Level     | Handle             |
++--------------------------+-------+--------------------+--------------------+
```
Add Tx Power Level as a return parameter. If the controller does not support the requested TX power, then the actual one will be in the Command Complete return value. For example, if you request -50dBm on a Nordic chip, the closest one supported is actually -40dBm, which would be what you return in the Command Complete.
Please also document that the controller will set the TX power to the closest supported level.

carlescufi replied:
See the  `7.8.53 LE Set Extended Advertising Parameters` in the spec and follow the format there. Please also split the parameters in 2 sections:

**Command parameters:**
and
**Return parameters**

just like the spec does. Name the return parameter as `Selected_Tx_Power` just like the spec does.

rstoica replied:
In general I have no problem following the spec convention and using the separation

**Command parameters:**
and 
**Return parameters:**

but then I will completely brake the formatting conventions of this entire file - see for instance the [Zephyr Set Trace Enable Command](https://github.com/zephyrproject-rtos/zephyr/pull/19779/files/32c04d9968f3c5cce18cad74a4cdb633c0a701ce..de505b96597559c418fbdf6def1fb7b6b91b3e68#diff-9c37bb9bf44e4a2035f47bde14315da6R447) and alike.

So I'd rather keep the original formatting of this file and if-need-be the Command/Return separation could be handled in a separate commit. The naming conventions you and @joerchan proposed will be of course addressed.

joerchan replied:
> but then I will completely brake the formatting conventions of this entire file - see for instance the Zephyr Set Trace Enable Command and alike.

Fair point. Follow the current convention. We can update to the spec convention for all commands as a separate fix.

rstoica replied:
Added the `Selected_Tx_Power` [return parameter](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-d32c7d39d131dcc56d85c8b8b7d5b328R148) which is yielded by each radio upon its Tx power flooring operation. The [documentation](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/6012d2b2ca1f5515ee0ae6dfc760301bfc4b0faa#diff-9c37bb9bf44e4a2035f47bde14315da6R908) reflects that as well.

carlescufi commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -800,6 +800,140 @@ When the Set_Scan_Request_Reports command has completed, a Command Complete
+| Command                  | OCF   | Command            | Return             |
+|                          |       | Parameters         | Parameters         |
++--------------------------+-------+--------------------+--------------------+
+| Write_Tx_Power_Level     | 0x00E | LL_State,          | Status,            |
```
Perhaps stay very vague and just call it `TX_Type`?

rstoica replied:
The resolution was to use `Handle_Type` and `Handle` pair to identify LL links independent of role/state of the links. This is also providing some flexibility regarding AE support and also provides a good model for other prospective Zephyr HCI VS commands.

The changes are visible [here](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-d32c7d39d131dcc56d85c8b8b7d5b328R133) and reflected also in the [docs](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/6012d2b2ca1f5515ee0ae6dfc760301bfc4b0faa#diff-9c37bb9bf44e4a2035f47bde14315da6R803)

carlescufi commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -800,6 +800,140 @@ When the Set_Scan_Request_Reports command has completed, a Command Complete
+	| 0x00               | Advertising LL state                 |
+	| 0x01               | Scanning LL state                    |
+	| 0x02               | Connected LL state                   |
+	| 0x03-0xFF          | Reserved (leads to HCI command error)|
```
remove the `(leads to ...)`

carlescufi commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -800,6 +800,140 @@ When the Set_Scan_Request_Reports command has completed, a Command Complete
+	+--------------------+--------------------------------------+
+	| Value              | Parameter Description                |
+	+--------------------+--------------------------------------+
+	| 0x0000 - 0xFFFF    | Handle ID of the state instance to   |
```
Please be specific here about what this parameter means for each `TX_Type`. For scanning, it is ignored. For advertising, it's either ignored (no Advertising Extensions) or maps to Advertising Handle (0x00 to 0xEF). For connections, it's Connection Handle (0x0 to 0xEFF).

carlescufi commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -800,6 +800,140 @@ When the Set_Scan_Request_Reports command has completed, a Command Complete
+	+--------------------+--------------------------------------+
+	| Value              | Parameter Description                |
+	+--------------------+--------------------------------------+
+	| 0x0000 - 0xFFFF    | Handle ID of the state instance to   |
```
Please be specific here about what this parameter means for each TX_Type. For scanning, it is ignored. For advertising, it's either ignored (no Advertising Extensions) or maps to Advertising Handle (0x00 to 0xEF). For connections, it's Connection Handle (0x0 to 0xEFF).

rstoica replied:
As discussed addressed this both in code for [advertiser](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-611c010769adb02c1f8eb406479efa93R147) and [scanner](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-611c010769adb02c1f8eb406479efa93R171) and also in [docs](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/6012d2b2ca1f5515ee0ae6dfc760301bfc4b0faa#diff-9c37bb9bf44e4a2035f47bde14315da6R853).

carlescufi commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -800,6 +800,140 @@ When the Set_Scan_Request_Reports command has completed, a Command Complete
+| Read_Tx_Power_Level      | 0x00F | LL_State,          | Status,            |
+|                          |       | Handle             | LL_State,          |
+|                          |       |                    | Handle,            |
+|                          |       |                    | Tx Power Level     |
```
Agreed. Please use the exact same formatting as the spec does in general.

rstoica replied:
Reformatted the docs as close as possible to the Core BT spec without breaking the existing Zephyr HCI [docs](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/6012d2b2ca1f5515ee0ae6dfc760301bfc4b0faa#diff-9c37bb9bf44e4a2035f47bde14315da6R803) formatting.

joerchan commented:
> To align the 2 HCI commands would basically imply the +1 shift in the role mapping 

Actually, in HCI valid values usually start from 0, so I think we should continue to do that, so skip that change.

> Shall the naming convention also be followed? I believe so, especially for the role part, whereas for the Tx power level I think your previous suggestion, i.e. Tx_Power_level is better suited following the Core BT spec.

We should follow the naming of the Core BT spec. The role part is the one thing we still haven't been able to agree on. I still think `handle_type` makes the most sense. But `role_type` would also work. I think we should try to avoid tying it  specifically to tx_power in case we would like to repeat the pattern.

> I am not sure about the power levels. Where would they go? In hci_vs.h seems a bit forced since they are vendor specific. Would everyone be supposed to add their own ones there

I think there can be a big difference in the amount of TX power different controllers support, so the best thing would be to do as carles already mentioned elsewhere to have selected TX power as output of the command as well. That way you can just specify a TX power level that you want , and the controller will select the closest thing it support. Because getting "invalid parameter" for that would be really annoying.
In this way we don't need to depend on the specific TX power levels that the controller support. And we don't have to put them anywhere.

You have a lot of comments to deal with now. But I think we have an agreement on how to proceed among the people involved. But in order to resolve this in an efficient manner please contact us on slack for any questions you might have :)

rstoica commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -289,6 +289,14 @@ config BT_CTLR_TX_PWR_MINUS_40
 
 endchoice
 
+config BT_CTRL_TX_PWR_DYNAMIC_CTRL
```
Rename this to `BT_CTLR_TX_PWR_DYNAMIC_CONTROL` as is more verbose and avoids the `CTRL` vs `CTLR` confusion.

`BT_CTLR` is the default Zephyr Bluetooth controller Kconfig abbreviation.

rstoica replied:
Solved [here](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-fdf68fcfb91aad56ca9d8b0a627c1ab4R292)

rstoica commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -7,6 +7,12 @@
 
 #define LL_VERSION_NUMBER BT_HCI_VERSION_5_1
 
+enum ll_txp {
```
As discussed this struct has been renamed as `struct bt_hci_vs_handle_type` [here](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/b8c7a15362e46dbcc1238e7bf79694779119a062#diff-65c5dce42a06bd585801995fc7a03ad0R10) so that the pattern could be reused by possible future HCI VS commands that need to implement functionality through the lower layers.

joerchan commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -800,6 +800,221 @@ When the Set_Scan_Request_Reports command has completed, a Command Complete
+	| -127 <= N <= 126   | The controller selected Tx power to  |
+	|                    | set for the given Handle_Type and    |
+	|                    | and pair formatted as a 1 octet      |
+	|                    | signed integer. The controller shall |
```
"The controller shall not modify this transmit power unless it is directed to do so by the host"
Does this include the No preference value?
I'm assuming that if the host set "no preference" then the controller is free to do dynamic TX power adjustments and not have to commit to the value it returned in this case?

rstoica replied:
No, it does not include the "No preference" value.

If host has no preference than the controller will revert to using the default Tx power as [mentioned](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/6012d2b2ca1f5515ee0ae6dfc760301bfc4b0faa#diff-9c37bb9bf44e4a2035f47bde14315da6R897).

I used the formatting and wording model from `7.8.53 LE Set Extended Advertising Parameters Command` in the Core Spec.

rstoica replied:
I will mark this as resolved for now as I don't see any issue here as the best next thing the controller can do if host has no preference is to actually use the defaults :)

joerchan commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,318 @@
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_VS_WRITE_TX_POWER_LEVEL,
+				   buf, &rsp);
+	if (err) {
+		printk("TXP SET HCI err: %d\n", err);
```
```suggestion
		printk("TXP SET HCI err: %d\n", err);
		return;
```

rstoica replied:
Fixed for all HCI, i.e. [setter](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/787709a21828771d56ba998996e9a82646846617#diff-b9bd630d9245e9d01d535cbaa393a233R106), [getter](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/787709a21828771d56ba998996e9a82646846617#diff-b9bd630d9245e9d01d535cbaa393a233R143) and [rssi](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/787709a21828771d56ba998996e9a82646846617#diff-b9bd630d9245e9d01d535cbaa393a233R69), again forgot about the cmd_send buffer handling.

joerchan commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,318 @@
+	bt_gatt_hrs_notify(heartrate);
+}
+
+void modulate_adv_tx_power(void *p1, void *p2, void *p3)
```
nitpick: it modulates for adv and connections, maybe remove _adv_?

rstoica replied:
[Solved](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/787709a21828771d56ba998996e9a82646846617#diff-b9bd630d9245e9d01d535cbaa393a233R235)

rstoica replied:
[Fixed](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/787709a21828771d56ba998996e9a82646846617#diff-b9bd630d9245e9d01d535cbaa393a233R235)

joerchan commented on samples/bluetooth/hci_pwr_ctrl/src/main.c:
```diff
@@ -0,0 +1,318 @@
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_READ_RSSI, buf, &rsp);
+	if (err) {
+		printk("RSSI HCI err: %d\n", err);
```
This is not HCI error. Also `bt_hci_cmd_send_sync` can fail before receiving the rsp.
So we can't safely use rsp if err is received.
I would suggest to do this:
```
struct net_buf *buf, *rsp = NULL;
...
if (err) {
	u8_t reason = rsp ? ((struct bt_hci_rp_read_rssi *)rsp->data)->status : 0;
	printk("Read RSSI err: %d\n reason 0x%02x", err, status);
	return;
}
```
Needs to be handled for the other cases where `rsp` is used in this way as well.

rstoica replied:
Thanks for pointing it out! I see why now. I saw now that `bt_hci_cmd_send_sync` may fail without allocation of `*rsp`. 

I implemented the changes throughout in all the functions handling the HCI calls. Here the changes for the `read_conn_rssi` - [pointer initialization](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/c8f060e0436bb073728d128cfba438c930a1f323#diff-b9bd630d9245e9d01d535cbaa393a233R47) and [exit logic](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/c8f060e0436bb073728d128cfba438c930a1f323#diff-b9bd630d9245e9d01d535cbaa393a233R63-R68)

joerchan reviewed (changes requested):
Possible use of invalid pointer.

mfiumara commented:
Is this ready to be merged?

joerchan commented:
> Is this ready to be merged?

There is currently feature-freeze on master branch. So it will have to wait until that is finished atleast.

cvinayak commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -7,6 +7,12 @@
 
 #define LL_VERSION_NUMBER BT_HCI_VERSION_5_1
 
+enum bt_hci_vs_handle_type {
```
Why redefine the handle type again here when the values are in hci_vs.h ? Drop this, and use u8_t in ll_tx_pwr_lvl_set/get.

rstoica replied:
The main reason behind this was, @joerchan please correct me if I am wrong in any way, to allow others to use this pattern in the LL split for other implementations, whereas the definitions in the hci_vs.h particular to this implementation where provide for the HCI interface of the set/get tx power level commands.

joerchan replied:
I think I wanted the enum to be independant of TX power, so that it could be re-used for similar features that can be configured for all LL roles. Redefining the enum is probably not wise, so should include `hci_vs.h` instead and use the definition from there. If u8_t is desired as the argument type then it is probably best to follow that advice. But need to include the `hci_vs.h` header for the source file where the switch on the enum values are.

rstoica replied:
Ok, @joerchan, I understand. To resolve both of your request and @cvinayak's, I suggest the following:

- remove `enum bt_hci_vs_handle_type` completely and use `u8_t` instead
- use macro define values from `hci_vs.h` and thus indeed include  `hci_vs.h` where ever there are checks on the type of the handle
- rename macro define names to make them more generic ? and thus not tie them to the power control and encourage the pattern for other purposes (here the idea is to change the name from `BT_HCI_VS_LL_TX_POWER_CONTROL_*` to `BT_HCI_VS_LL_HANDLE_TYPE_*` to make the enum independent of the TX power)

Would these be then okay for both ?

rstoica replied:
Fixed as discussed - [`hci_vs.h`](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/856709c837c8feb553e54c4b84fc44174136897f#diff-d32c7d39d131dcc56d85c8b8b7d5b328R133) and [`ll.h`](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/856709c837c8feb553e54c4b84fc44174136897f#diff-65c5dce42a06bd585801995fc7a03ad0R84)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -21,6 +21,10 @@
 #include "ll_feat.h"
 #include "ll_settings.h"
 
+#if defined(CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL)
```
Dont need the conditional compile unless the includes are not implicitly excluded by compiler when not being referenced in following code in this file.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_tx_pwr.c:
```diff
@@ -17,36 +17,198 @@
 
-#if defined(CONFIG_BT_LL_SW_SPLIT)
-u8_t ll_tx_pwr_lvl_get(u16_t handle, u8_t type, s8_t *tx_pwr_lvl)
+#if defined(CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL)
```
This too can be removed?

cvinayak reviewed (approved):
LGTM. Approving this, pending testing on my end, in case fixes required, can be done subsequently as new fix PR.

carlescufi commented:
@jhedberg @thoh-ot Any objections to merging this?

kelsayed commented:
Hello, 

I am working on the power control feature of Core Spec 5.2. I am almost 60% defining the air-messages and the state machine within the llcp and he actions/behavior upon receiving PWR_CTRL_REQ/RSP or a related REJ. Still untested but it is in the pipeline. 
Main issues now are:
1) Handling the APR (acceptable power reduction) in the PWR_CTRL_RSP
2) Insuring there are no leaks or any PWR_CTRL received pdu not released
3) Issuing command to change PHY TX power (I used the host based TX power that this PR is addressing which seems to only change the connection TX power in the connection record and I guess the radio manager will then somehow use that to control the PHY?)
4) Overall testing. I need help here from the Zephyr Bluetooth developers. 

I will mail this to the zephyr development list shortly. Should I start a PR or is there any on-going effort for BLE core spec 5.2 implementation that I may join forces with?

rstoica commented:
> Hello,
> 
> I am working on the power control feature of Core Spec 5.2. I am almost 60% defining the air-messages and the state machine within the llcp and he actions/behavior upon receiving PWR_CTRL_REQ/RSP or a related REJ. Still untested but it is in the pipeline.
> Main issues now are:
> 
> 1. Handling the APR (acceptable power reduction) in the PWR_CTRL_RSP
> 2. Insuring there are no leaks or any PWR_CTRL received pdu not released
> 3. Issuing command to change PHY TX power (I used the host based TX power that this PR is addressing which seems to only change the connection TX power in the connection record and I guess the radio manager will then somehow use that to control the PHY?)
> 4. Overall testing. I need help here from the Zephyr Bluetooth developers.
> 
> I will mail this to the zephyr development list shortly. Should I start a PR or is there any on-going effort for BLE core spec 5.2 implementation that I may join forces with?

Hi @kelsayed, thanks for your interest in Zephyr and its bluetooth stack. I think this is a topic that should be discussed elsewhere. I suggest using the [Zephyr slack workspace](https://docs.zephyrproject.org/latest/guides/getting-help.html) and reposting this into the #bluetooth channel. There you can get the feedback you need to continue in alignment with the planned development. Join the Slack workspace with this [link](https://tinyurl.com/y5glwylp)
